platform/qt_location_service.cpp

```diff
@@ -0,0 +1,172 @@
+
+#include "std/target_os.hpp"
+
+#include <QGeoPositionInfoSource>
+
+namespace {
+  static location::GpsInfo gpsInfoFromQGeoPositionInfo(const QGeoPositionInfo& i, location::TLocationSource source)
+  {
+    location::GpsInfo info;
+    info.m_source = source;
+
+    info.m_latitude = i.coordinate().latitude();
+    info.m_longitude = i.coordinate().longitude();
+    info.m_timestamp = (double)i.timestamp().toSecsSinceEpoch();
+
+    if (i.hasAttribute(QGeoPositionInfo::HorizontalAccuracy))
+    {
+      info.m_horizontalAccuracy = (double)i.attribute(QGeoPositionInfo::HorizontalAccuracy);
+    }
+
+    if (i.hasAttribute(QGeoPositionInfo::VerticalAccuracy))
+    {
+      info.m_verticalAccuracy = (double)i.attribute(QGeoPositionInfo::VerticalAccuracy);
+    }
+
+    if (i.hasAttribute(QGeoPositionInfo::Direction))
+    {
+      info.m_bearing = (double)i.attribute(QGeoPositionInfo::Direction);
+    }
+
+    if (i.hasAttribute(QGeoPositionInfo::GroundSpeed))
+    {
+      info.m_speedMpS = (double)i.attribute(QGeoPositionInfo::GroundSpeed);
+    }
+
+    return info;
+  }
+
+  static location::TLocationError tLocationErrorFromQGeoPositionInfoError(QGeoPositionInfoSource::Error error)
+  {
+     location::TLocationError result;
+     switch(error)
+     {
+       case QGeoPositionInfoSource::AccessError: result = location::TLocationError::EDenied; break;
+       case QGeoPositionInfoSource::ClosedError: result = location::TLocationError::EGPSIsOff; break;
+       case QGeoPositionInfoSource::NoError: result = location::TLocationError::ENoError; break;
+       case QGeoPositionInfoSource::UnknownSourceError: [[fallthrough]];
+       case QGeoPositionInfoSource::UpdateTimeoutError: [[fallthrough]];
+       default: result = location::TLocationError::ENotSupported; break;
+     }
+     return result;
+  }
+
+  static location::TLocationSource qStringToTLocationSource(const QString& sourceName)
+  {
+      if ("geoclue2" == sourceName)
+      {
+          return location::TLocationSource::EGeoClue2;
+      }
+      return location::TLocationSource::EUndefined;
+  }
+};
+
+
+QtLocationService::QtLocationService(location::LocationObserver & observer, const std::string & sourceName) : LocationService(observer)
+{
+  QVariantMap params;
+  params["desktopId"] = "OrganicMaps";
+  m_positionSource = QGeoPositionInfoSource::createSource(QString::fromStdString(sourceName), params, this);
+
+  if (!m_positionSource)
+  {
+    LOG(LWARNING, ("Failed to acquire QGeoPositionInfoSource from ", sourceName));
+    return;
+  }
+
+  bool isConnected = (bool)connect(m_positionSource, &QGeoPositionInfoSource::positionUpdated,
+      this, &QtLocationService::OnLocationUpdate);
+
+  if (!isConnected)
+  {
+    LOG(LERROR, ("Failed to connect the signal:", "positionUpdated"));
+    return;
+  }
+
+  LOG(LDEBUG, ("Signal successfully connected:", "positionUpdated"));
+
+  isConnected = (bool) connect(m_positionSource, &QGeoPositionInfoSource::errorOccurred,
```

Comment:
Do you really need this isConnected variable? Won't this work?
```
if (!connect(...))
{
  CHECK(false, ("If the goal is to immediately crash so developer can fix the error."));
  // or
  LOG(LWARNING, ("Some non-critical, recoverable error"));
  return false;
}
```

Reply:
I don't need it. For me it is even better if I eliminate the `isConnected` , it was simply a courtesy to the reader.
And yes the goal is to immediately crash, because it is a sign of a developer error if it is not properly connected.

Reply:
Done